samplers/samplers.go

```diff
@@ -72,6 +74,101 @@ type JSONMetric struct {
 	Value []byte `json:"value"`
 }
 
+const CardinalityCountName = "cardinality.count"
+const CardinalityCountType = "cardinality_count"
+
+// CardinalityCount is a sampler that records an approximate cardinality count
+// by metric name.
+type CardinalityCount struct {
+	MetricCardinality map[string]*hyperloglog.HyperLogLogPlus
+}
+
+func NewCardinalityCount() *CardinalityCount {
+	return &CardinalityCount{make(map[string]*hyperloglog.HyperLogLogPlus)}
+}
+
+// Sample adds a measurement to the cardinality counter. It makes the
+// assumption that no two metrics with different types share a name (ie, name
+// is unique across all types)
+func (cc *CardinalityCount) Sample(name string, joinedTags string) {
+	if _, present := cc.MetricCardinality[name]; !present {
+		hll, _ := hyperloglog.NewPlus(18)
+		cc.MetricCardinality[name] = hll
+	}
+	hasher := fnv.New64a()
+	hasher.Write([]byte(joinedTags))
+	cc.MetricCardinality[name].Add(hasher)
+}
+
+// Export groups the CardinalityCount struct into sub structs, grouped by what
+// their metric consistently hashes to
+func (cc *CardinalityCount) ExportSets() ([]JSONMetric, error) {
+	jsonMetrics := make([]JSONMetric, 0, len(cc.MetricCardinality))
+	for metricName, hll := range cc.MetricCardinality {
+
+		buf := new(bytes.Buffer)
+		encoder := gob.NewEncoder(buf)
+
+		err := encoder.Encode(map[string]*hyperloglog.HyperLogLogPlus{metricName: hll})
+		if err != nil {
+			// TODO (kiran): do we return an array of metrics, instead?
+			log.Printf("failed to export cardinality count for metric name:%s, error:%v", metricName, err)
```

Comment:
At the moment, we don't use logrus (and therefore our Sentry hooks) here. Ideally, this should go to Sentry, though that'll require a little more work to configure, since this is in a subpackage.

Reply:
from IRL: we're going to make a custom error type to return to the caller here, and have the caller emit to Sentry via logrus, instead